acond/src/restful.rs

```diff
@@ -0,0 +1,577 @@
+    }
+}
+
+async fn inspect(
+    request: web::Query<InspectRequest>,
+    service: web::Data<ExchangeService>,
+) -> Result<HttpResponse, RestError> {
+    let mut request_buf: Vec<u8> =
+        bincode::serialize(&request.into_inner()).map_err(|e| RestError::Unknown(e.to_string()))?;
+    let response_buf = service
+        .send_recv(8, &mut request_buf)
+        .await
+        .map_err(|e| RestError::Unknown(e.to_string()))?;
+
+    match response_buf[0] {
+        0 => {
+            let response: InspectResponse = bincode::deserialize(&response_buf[1..])
+                .map_err(|e| RestError::Unknown(e.to_string()))?;
+            Ok(HttpResponse::Ok().json(response))
+        }
+        1 => Err(get_rest_error(&response_buf[1..]).map_err(|e| RestError::Unknown(e.to_string()))?),
+        _ => Err(RestError::Unknown("Server response format error".into())),
+    }
+}
+
+async fn report(
+    request: web::Query<ReportRequest>,
+    service: web::Data<ExchangeService>,
+) -> Result<HttpResponse, RestError> {
+    let mut request_buf: Vec<u8> =
+        bincode::serialize(&request.into_inner()).map_err(|e| RestError::Unknown(e.to_string()))?;
+    let response_buf = service
+        .send_recv(9, &mut request_buf)
+        .await
+        .map_err(|e| RestError::Unknown(e.to_string()))?;
+
+    match response_buf[0] {
+        0 => {
+            let response: ReportResponse = bincode::deserialize(&response_buf[1..])
+                .map_err(|e| RestError::Unknown(e.to_string()))?;
+            Ok(HttpResponse::Ok().json(response))
+        }
+        1 => Err(get_rest_error(&response_buf[1..]).map_err(|e| RestError::Unknown(e.to_string()))?),
+        _ => Err(RestError::Unknown("Server response format error".into())),
+    }
+}
+struct ExchangeService {
+    stream: Arc<Mutex<UnixStream>>,
+}
+
+impl ExchangeService {
+    fn new(stream: UnixStream) -> Self {
+        Self {
+            stream: Arc::new(Mutex::new(stream)),
+        }
+    }
+
+    async fn send_recv(&self, command: u8, buf: &mut Vec<u8>) -> Result<Vec<u8>> {
+        buf.insert(0, command);
+
+        let mut send_buf = (buf.len() as u32).to_ne_bytes().to_vec();
+        send_buf.append(buf);
+        acond_io::write_async_lock(self.stream.clone(), &send_buf, send_buf.len()).await?;
+
+        let recv_buf = acond_io::read_async_lock(self.stream.clone()).await?;
+        if recv_buf.is_empty() {
+            return Err(anyhow!(utils::ERR_UNEXPECTED));
+        }
+        Ok(recv_buf)
+    }
+
+    async fn send_recv2(&self, command: u8, buf: &mut Vec<u8>, path: &Path) -> Result<Vec<u8>> {
+        buf.insert(0, command);
+
+        let mut send_buf = (buf.len() as u32).to_ne_bytes().to_vec();
+        send_buf.append(buf);
+        acond_io::write_async_lock(self.stream.clone(), &send_buf, send_buf.len()).await?;
+        {
+            let ref_stream = self.stream.clone();
+            let stream = ref_stream.lock().await;
```

Comment:
Why is `clone()` necessary before `lock()`?

Reply:
This creates another pointer to the same allocation, increasing the strong reference count.